File: examples/php/webhook_selftest.php
<?php
// Runs the shared vectors (examples/testdata/webhook-vectors.json) through dona_verify_webhook.
declare(strict_types=1);

require __DIR__ . '/webhook.php';

$vectors = json_decode(file_get_contents(__DIR__ . '/../testdata/webhook-vectors.json'), true, flags: JSON_THROW_ON_ERROR);
$failed = 0;
foreach ($vectors['cases'] as $c) {
    $headers = ['webhook-id' => $c['id'], 'webhook-timestamp' => (string) $c['timestamp'], 'webhook-signature' => $c['signature']];
    try {
        dona_verify_webhook($c['secret'], $headers, $c['body'], $c['now']);
        $valid = true;
    } catch (DonaWebhookVerificationError) {
        $valid = false;
    }
    if ($valid !== $c['valid']) {
        $failed++;
        fwrite(STDERR, sprintf("FAIL %s: got valid=%s, want %s\n", $c['name'], var_export($valid, true), var_export($c['valid'], true)));
    }
}
printf("php webhook: %d/%d vectors\n", count($vectors['cases']) - $failed, count($vectors['cases']));
exit($failed ? 1 : 0);

File: examples/python/webhook_selftest.py
"""Runs the shared vectors (examples/testdata/webhook-vectors.json) through verify_webhook."""

import json
import pathlib
import sys

from webhook import WebhookVerificationError, verify_webhook

vectors = json.loads((pathlib.Path(__file__).parent.parent / "testdata" / "webhook-vectors.json").read_text("utf-8"))
failed = 0
for c in vectors["cases"]:
    headers = {"webhook-id": c["id"], "webhook-timestamp": str(c["timestamp"]), "webhook-signature": c["signature"]}
    try:
        verify_webhook(c["secret"], headers, c["body"], now=c["now"])
        valid = True
    except WebhookVerificationError:
        valid = False
    if valid != c["valid"]:
        failed += 1
        print(f"FAIL {c['name']}: got valid={valid}, want {c['valid']}", file=sys.stderr)
print(f"python webhook: {len(vectors['cases']) - failed}/{len(vectors['cases'])} vectors")
sys.exit(1 if failed else 0)
